Feat/create blacklist screen

hellypi opened the pull request:
Add blacklist page so user can blacklist specific cryptocurrency addresses.

hellypi committed:
feat: Blacklist page

hellypi committed:
feat: Blacklist page

chaotixkilla committed:
feat: token management screen

chaotixkilla committed:
feat: add revoke token button

RafaelTaranto commented on lib/blacklistQueries.js:
```diff
@@ -0,0 +1,52 @@
+const db = require('./db')
```
There's no reason to rename this file. You can revert the name back to blacklist.js

RafaelTaranto commented on lib/cash-in/cash-in-tx.js:
```diff
@@ -2,7 +2,7 @@ const _ = require('lodash/fp')
 const pgp = require('pg-promise')()
```
This whole file was changed only because of automatic formatting that's not really configured and of the blacklist file rename. You can undo all changes here.

RafaelTaranto commented on lib/new-admin/graphql/schema.js:
```diff
@@ -222,18 +247,28 @@ const typeDefs = gql`
     saveConfig(config: JSONObject): JSONObject
     createPairingTotem(name: String!): String
     saveAccounts(accounts: JSONObject): JSONObject
+    deleteBlacklistRow(cryptoCode: String, address: String): [String]
```
Don't we want cryptoCode and address to be required here as well? Is there are reason why this returns an array?

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.js:
```diff
@@ -0,0 +1,141 @@
+          />
+        </div>
+      </Grid>
+      <BlackListModal
+        showModal={showModal}
+        toggleModal={toggleModal}
+        selectedCoin={clickedItem}
+        addToBlacklist={addToBlacklist}></BlackListModal>
```
This can be self closing

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.js:
```diff
@@ -0,0 +1,141 @@
+const Blacklist = () => {
+  const { data: blacklistResponse } = useQuery(GET_BLACKLIST)
+  const { data: configData } = useQuery(GET_INFO)
+  const blacklistData = R.path(['blacklist'])(blacklistResponse) ?? []
+  const availableCurrencies =
+    R.path(['cryptoCurrencies'], blacklistResponse) ?? []
+
```
Hooks have a bunch of rules for them to work properly, because of this I tend to prefer to get group them on the top of the function. 

Try to move all hooks to the top and the logic related to them afterwards.

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.js:
```diff
@@ -0,0 +1,141 @@
+    display: 'Bitcoin'
+  })
+
+  const groupByCode = R.groupBy(obj => obj.cryptoCode)
```
Another thing that helps with readability is moving pure functions out of the component, this one can be outside the Blacklist function definition.

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.js:
```diff
@@ -0,0 +1,141 @@
+    refetchQueries: () => ['getData']
+  })
+
+  const isSelected = it => {
+    return clickedItem.code === it.code
+  }
```
you can remove the brackets and the return to shorten this up

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.js:
```diff
@@ -0,0 +1,141 @@
+    return clickedItem.code === it.code
+  }
+
+  const toggleModal = () => {
+    setShowModal(!showModal)
+  }
```
you can remove brackets here as well

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.js:
```diff
@@ -0,0 +1,141 @@
+            data={formattedData}
+            selectedCoin={clickedItem}
+            handleDeleteEntry={handleDeleteEntry}
+            saveConfig={saveConfig}
+            configData={configData}
```
We can probably simplify this into moving the logic to the outside of the BlacklistTable component. The idea is to make the component as dumb as possible with relation to the context. 

Instead of passing down the configData as a whole you can just pass down rejectAddressReuse and have Blacklist.js figure out the namespacing logic, the same is true for saving.

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js:
```diff
@@ -0,0 +1,42 @@
+    display: 'flex',
+    flexDirection: 'column',
+    flex: 1,
+    marginLeft: 48
```
We're not really using this everywhere but it's good practice to use our `spacer` variable. You can import it from 'src/styling/variables'. 

Then you can make it spacer * 6 instead of 48.

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js:
```diff
@@ -0,0 +1,42 @@
+    flex: 1,
+    marginLeft: 48
+  },
+  bottomRight: {
```
This can do with a different name. Like `footer`.

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js:
```diff
@@ -0,0 +1,42 @@
+    marginLeft: 48
+  },
+  bottomRight: {
+    marginLeft: 'auto',
+    marginTop: 'auto',
+    marginBottom: '24px'
```
this can be 
`margin: [['auto', 0, spacer * 3, 'auto']]`

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js:
```diff
@@ -0,0 +1,42 @@
+    marginTop: 'auto',
+    marginBottom: '24px'
+  },
+  modalTitle: {
+    lineHeight: '120%',
+    color: '#1b2559',
+    fontSize: 14,
+    fontFamily: 'Mont',
+    fontWeight: 900
+  },
```
Isn't this already defined in typography? We don't want to manually set font styles nor use unnamed colors. The colors are all exported with names from the styling/variables file.

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js:
```diff
@@ -0,0 +1,42 @@
+    flexDirection: 'row'
+  },
+  white: {
+    color: 'white'
```
even though it shouldn't be a big deal, we're exporting white as a named variable as well. The reasoning for naming everything instead of using a const is that if we decide to slightly change our base white color the change will be applied to all places if we're using the variable name.

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js:
```diff
@@ -0,0 +1,42 @@
+    color: 'white'
+  },
+  deleteButton: {
+    paddingLeft: '13px'
```
You can shorten this to

`paddingLeft: 13` you can use numbers instead of strings with `px` at the end.

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js:
```diff
@@ -0,0 +1,42 @@
+    paddingLeft: '13px'
+  },
+  addressRow: {
+    marginLeft: '8px'
```
same thing as above

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js:
```diff
@@ -0,0 +1,42 @@
+    marginLeft: '8px'
+  },
+  table: {
+    height: '900px'
```
Any reason for having this specified height on table? We probably want to to follow the size of the page.

RafaelTaranto commented on new-lamassu-admin/src/routing/routes.js:
```diff
@@ -10,6 +10,8 @@ import {
 
 import { AppContext } from 'src/App'
 import AuthRegister from 'src/pages/AuthRegister'
+// import should appear before cashout (why?)
```
You can remove this comment. I think I already explained this in a call, but imports are organized by scope and alphabetically.

RafaelTaranto commented on new-lamassu-admin/src/routing/routes.js:
```diff
@@ -147,6 +149,13 @@ const tree = [
+        route: '/compliance/blacklist',
+        component: Blacklist
+      },
+      // what is this last one for
```
What is this comment?

hellypi replied:
That was a question I wanted to ask you but ended up forgetting. I started taking notes on the side instead of commenting, that one escaped my eye...

liordino committed:
fix: prompt when leaving page without saving changes on editable tables

liordino committed:
fix: keep previous value when closing the configuration popup without
saving on locales screen

refactor: renamed enableCoin function to configureCoin

fix: keep previous value when closing the configuration popup without
saving on wallets screen

refactor: renamed enableThirdPartyService function to
configureThirdPartyService

chaotixkilla committed:
feat: add request data to user_tokens and token filtering

chaotixkilla committed:
refactor: user agent parsing

hellypi commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Create the blacklist screen](https://trello.com/c/OjxjUUH7/165-create-the-blacklist-screen)

csrapr commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Create the blacklist screen](https://trello.com/c/OjxjUUH7/165-create-the-blacklist-screen)

hellypi committed:
fix: Brings unrelated logic outside of BlacklistTable.js

hellypi committed:
fix: Brings unrelated logic outside of BlacklistTable.js

hellypi committed:
fix: Revert changes to lib/cash-in/cash-in-tx.js

hellypi committed:
fix: Revert changes to lib/cash-in/cash-in-tx.js

hellypi committed:
fix: fixed mismatched type on delete blacklist row query

hellypi committed:
fix: fixed mismatched type on delete blacklist row query

RafaelTaranto committed:
feat: error handling on forms

liordino committed:
fix: removed 'Share with Lamassu' button from machine logs and server
maintanence pages

liordino committed:
feat: added the transaction status on the transactions table

liordino committed:
fix: wrong property was being used to determine if a tx was cash-in or
cash-out on the transactions page, resulting in wrong cash-out status on
the page

liordino committed:
fix: renamed column headers to 'Cassete' instead of 'Cashout'

fix: cassete denominations were inverted

liordino committed:
fix: replaced country info with state/province and added birth date on
the id data card

liordino committed:
chore: removed unused variable from ComplianceDetails

fix: fixed the customer mutation, which was causing errors on all
customer compliance buttons

liordino committed:
fix: removed unnecessary support log code from server

liordino committed:
style: set IconButton stroke colors to comet

fix: replaced edit buttons of the BooleanPropertiesTable component and
Contact Info and Terms & Conditions pages

liordino committed:
fix: replaced 1 digit with a 2 digit day format in dates

liordino committed:
fix: refetch cassetes data to avoid resetting to the initial value on UI
when canceling edit

liordino committed:
feat: added shutdown function to the machine loader

feat: created shutdown route and call it from gql server (called from
admin UI button)

feat: added an extra message property to the ConfirmDialog

refactor: simplified the MachineDetailsCard component

feat: added an extra information message to the Shutdown machine action

RafaelTaranto committed:
feat: error handling for the tables and wizards

RafaelTaranto committed:
fix: receipt printing new admin

RafaelTaranto committed:
fix: server logs downloads

RafaelTaranto committed:
feat: return logs in csv format

RafaelTaranto committed:
fix: overrides should not merge child arrays

liordino committed:
fix: high value notifications, if configured, are now triggered even if
transactions notifications are globally disabled (for sms or email)

fix: added extra information on the sms notification informing if the
transaction that just happened is a high value one

liordino committed:
fix: removed the attempts to use the bignumber.js lib with null values

RafaelTaranto committed:
feat: nodemon and fast reload

liordino committed:
fix: correctly test the alert thresholds (considering falsy values)

liordino committed:
fix: move number alongside bar on empty part until 51%, and them move it
to the filled part

RafaelTaranto committed:
fix: unresponsive instead of stuck if both

RafaelTaranto committed:
chore: version fonts on main project

RafaelTaranto committed:
feat: all coins on commissions

RafaelTaranto committed:
fix: hide navbar before user authenticates

CrypticaScriptura committed:
chore: update wallet nodes

RafaelTaranto committed:
chore: version builded admin

RafaelTaranto committed:
Chore/merge dd back (#503)

* chore: update ZEC, LTC, BTC, ETH, use BCHN

* chore: update zcashd

* chore: v7.4.10

* fix: remove lamassu-update

* fix: remove lamassu-update from bins

* fix: use p2sh-segwit addresses for cash-out

Latest releases of bitcoind default to bech32 addresses. Instead, use p2sh-segwit to maintain compatibility on cash-out.

* fix: use p2sh-segwit addresses for LTC

* chore: update wallet nodes

Co-authored-by: naconner <7396812+naconner@users.noreply.github.com>
Co-authored-by: Neal <neal@lamassu.is>

RafaelTaranto committed:
fix: wizard locales regression (#504)

RafaelTaranto committed:
Fix/csv logs (#505)

* fix: use async csv lib

* fix: wording on the wizard

* chore: builded react files

RafaelTaranto commented on lib/blacklist.js:
```diff
@@ -1,22 +1,52 @@
+    })
+}
+
+function blocked(address, cryptoCode) {
```
Hmm, this doesn't pass standard. Functions should have a space between the name and parenthesis on function definition.

RafaelTaranto commented on lib/blacklist.js:
```diff
@@ -1,22 +1,52 @@
 }
 
-function addToUsedAddresses (address, cryptoCode) {
+function addToUsedAddresses(address, cryptoCode) {
```
Same thing here

RafaelTaranto commented on lib/cash-in/cash-in-tx.js:
```diff
@@ -176,4 +176,4 @@ function cancel (txId) {
     })
     .then(() => logActionById('operatorCompleted', {}, txId))
-}
+}
```
Undo the change of this file.

RafaelTaranto commented on new-lamassu-admin/src/utils/coins/bch.js:
```diff
@@ -0,0 +1,15 @@
+const cashaddr = require('cashaddrjs')
```
I don't see this added on package.json, this won't really work.

It might be too annoying to fix or move to the server side for all the addresses, remove all validation code.

We can circle back on this if the need arises.

hellypi replied:
all validation code for all coins or just for BCH?

RafaelTaranto commented on new-lamassu-admin/src/pages/Triggers/Triggers.js:
```diff
@@ -38,7 +38,7 @@ const Triggers = () => {
   const { data, loading } = useQuery(GET_INFO)
   const triggers = fromServer(data?.config?.triggers ?? [])
-
+  console.log(triggers)
```
undo changes to this file

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/BlacklistTable.js:
```diff
@@ -0,0 +1,62 @@
+    : data[R.keys(data)[0]]
+
+  return (
+    <>
```
If you're returning just a single node you don't really need the `<> </>`

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js:
```diff
@@ -0,0 +1,99 @@
+              ? `Blacklist ${R.toLower(selectedCoin.display)} address`
+              : ''}
+          </H3>
+          <TextInput
+            error={invalidAddress}
+            label="Paste new address to blacklist here"
+            name="address-to-block-input"
+            autoFocus
+            id="address-to-block-input"
+            type="text"
+            size="sm"
+            fullWidth
+            InputLabelProps={{ shrink: true }}
+            placeholder={`ex: ${placeholderAddress[selectedCoin.code]}`} // "ex: 0x309abbd2f85ead2fcbb5323e963550c88c8a569aca4e088e9020a03fd04bf4bd"
+            onChange={handleChange}
+            value={addressField}
+          />
```
Change this to use formik, you can create use simple validation since we're removing the coin specific parts.

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js:
```diff
@@ -0,0 +1,99 @@
+    BCH: 'qrd6za97wm03lfyg82w0c9vqgc727rhemg5yd9k3dm'
+  }
+
+  coinValidator.BTC('HELLO?')
```
Remove this

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js:
```diff
@@ -0,0 +1,99 @@
+}) => {
+  const classes = useStyles()
+
+  const [addressField, setAddressField] = useState('')
+  const [invalidAddress, setInvalidAddress] = useState(false)
+
+  const handleChange = event => {
+    if (event.target.value === '') {
+      setInvalidAddress(false)
+    }
+    setAddressField(event.target.value)
+  }
```
Formik should handle both validation code and field state

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js:
```diff
@@ -0,0 +1,99 @@
+    setAddressField(event.target.value)
+  }
+
+  const handleAddToBlacklist = () => {
+    if (addressField.trim() === '') {
+      setInvalidAddress(true)
+      return
+    }
+    if (coinValidator[selectedCoin.code](addressField.trim())) {
+      addToBlacklist(selectedCoin.code, addressField.trim())
+      handleClose()
+    } else {
+      setInvalidAddress(true)
+    }
+  }
```
We don't really wanna close the modal before getting the response from the server. See how other wizards handle this.

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js:
```diff
@@ -0,0 +1,99 @@
+    }
+  }
+
+  const handleClose = () => {
+    setAddressField('')
+    setInvalidAddress(false)
+    toggleModal()
+  }
```
State should be handled by formik and handleClose should call a external close handler.

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.js:
```diff
@@ -0,0 +1,183 @@
+  const { data: blacklistResponse } = useQuery(GET_BLACKLIST)
+  const { data: configData } = useQuery(GET_INFO)
+  const [showModal, setShowModal] = useState(false)
+  const [clickedItem, setClickedItem] = useState({
+    code: 'BTC',
+    display: 'Bitcoin'
+  })
```
Easier to just set the code as state, no reason to have a object here.

hellypi replied:
as we talked in private message, we do need this after all 👍 :)

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.js:
```diff
@@ -0,0 +1,183 @@
+      <Grid container className={classes.grid}>
+        <Sidebar
+          data={availableCurrencies}
+          isSelected={isSelected}
```
You can probably do R.propEq here instead of defining a isSelected function

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js:
```diff
@@ -0,0 +1,99 @@
+
+  return (
+    <>
+      {showModal && (
```
Better to have showModal handled on the outside component

RafaelTaranto commented on new-lamassu-admin/src/pages/Blacklist/Blacklist.js:
```diff
@@ -0,0 +1,183 @@
+      </Grid>
+      <BlackListModal
+        showModal={showModal}
+        toggleModal={toggleModal}
```
Name this onClose or similar, we don't need to ever toggle the modal inside the modal, pass down the `() => setShowModal(false)` to this

hellypi committed:
fix: Fixed pull request problems except formik related

hellypi committed:
fix: Fixed pull request problems except formik related

hellypi committed:
Merge branch 'feat/create-blacklist-screen' of github.com:csrapr/lamassu-server into feat/create-blacklist-screen

hellypi committed:
fix: Changed blacklistmodal input field to use formik
